test/StreamJsonRpc.Tests/MarshalableProxyTests.cs

```diff
@@ -0,0 +1,522 @@
+        clientRpc.TraceSource = new TraceSource("Client", SourceLevels.Verbose);
+
+        serverRpc.TraceSource.Listeners.Add(new XunitTraceListener(this.Logger));
+        clientRpc.TraceSource.Listeners.Add(new XunitTraceListener(this.Logger));
+
+        serverRpc.StartListening();
+
+        Data data = new Data();
+        await client.AcceptProxyAsync(data, dispose: false);
+
+        Assert.False(data.IsDisposed);
+
+        pipes.Item1.AsStream().Dispose();
+        await serverRpc.Completion.WithCancellation(this.TimeoutToken);
+
+        Assert.True(data.IsDisposed);
+    }
+
+    protected abstract IJsonRpcMessageFormatter CreateFormatter();
+
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    private async Task<(WeakReference Proxy, WeakReference Target)> DisposableReturnValue_Helper()
+    {
+        IDisposable? proxyDisposable = await this.client.GetMarshalableAsync();
+        Assert.NotNull(proxyDisposable);
+        Assumes.NotNull(this.server.ReturnedMarshalable);
+        Assert.False(this.server.ReturnedMarshalable.IsDisposed);
+        proxyDisposable!.Dispose();
+        WeakReference weakProxy = new WeakReference(proxyDisposable);
+
+        await this.server.ReturnedMarshalableDisposed.WaitAsync(this.TimeoutToken);
+        WeakReference weakTarget = new WeakReference(this.server.ReturnedMarshalable);
+        this.server.ReturnedMarshalable = null;
+        return (weakProxy, weakTarget);
+    }
+
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    private async Task<(WeakReference Proxy, WeakReference Target)> DisposableArg_Helper()
+    {
+        var disposed = new AsyncManualResetEvent();
+        var strongTarget = new Data(disposed.Set);
+        WeakReference weakTarget = new WeakReference(strongTarget);
+
+        await this.client.AcceptProxyAsync(strongTarget);
+        await disposed.WaitAsync(this.TimeoutToken);
+        Assumes.NotNull(this.server.ReceivedProxy);
+
+        WeakReference weakProxy = new WeakReference(this.server.ReceivedProxy);
+        this.server.ReceivedProxy = null;
+        return (weakProxy, weakTarget);
+    }
+
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    private async Task<(WeakReference Proxy, WeakReference Target)> DisposableWithinArg_Helper()
+    {
+        var disposed = new AsyncManualResetEvent();
+        var strongTarget = new Data(disposed.Set);
+        WeakReference weakTarget = new WeakReference(strongTarget);
+
+        await this.client.AcceptProxyContainerAsync(new ProxyContainer<IMarshalable> { Marshalable = strongTarget });
+        await disposed.WaitAsync(this.TimeoutToken);
+        Assumes.NotNull(this.server.ReceivedProxy);
+
+        WeakReference weakProxy = new WeakReference(this.server.ReceivedProxy);
+        this.server.ReceivedProxy = null;
+        return (weakProxy, weakTarget);
+    }
+
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    private async Task<WeakReference> NoLeakWhenTransmissionFailsAfterTokenGenerated_Helper()
```

Comment:
Such helper methods *do* need to be separate methods, but now that C# allows us to attribute local methods, the most maintainable and reader-friendly way to lay this out is to move all the helper methods into their test methods as local methods (typically named something shorter as well, like `Helper`).